### PR TITLE
Convert BaseModel to the new API

### PR DESCRIPTION
**Pull request by rco-odoo**

### CODE REVIEW

**Comment by rco-odoo** [comment]
@antonylesuisse @odony @KangOl you are welcome to review this migration.

**Comment by rco-odoo** [comment]
Functionally tested by @bst-odoo with success :+1:

**Comment by rco-odoo** [comment]
@xmo-odoo you are welcome to review this one.

File: openerp/osv/expression.py [hunk 1]
```diff
@@ -703,47 +704,51 @@ def to_ids(value, comodel, context=None, limit=None):
...
                 return [value]
             if names:
-                name_get_list = [name_get[0] for name in names for name_get in comodel.name_search(cr, uid, name, [], 'ilike', context=context, limit=limit)]
-                return list(set(name_get_list))
+                ids = [
+                    rid
+                    for name in names
+                    for rid, rname in comodel.name_search(name, [], 'ilike', limit=None)
+                ]
+                return list(set(ids))
```
Comment (xmo-odoo) [inline openerp/osv/expression.py]: use a set comprehension and listify it afterwards?

File: openerp/osv/expression.py [hunk 2]
```diff
@@ -703,47 +704,51 @@ def to_ids(value, comodel, context=None, limit=None):
...
-                    return [(left, 'in', left_model.search(cr, uid, doms, context=context))]
+                    return [(left, 'in', left_model.search(doms))]
                 return doms
             else:
                 def recursive_children(ids, model, parent_field):
                     if not ids:
                         return []
-                    ids2 = model.search(cr, uid, [(parent_field, 'in', ids)], context=context)
+                    ids2 = model.search([(parent_field, 'in', ids)]).ids
                     return ids + recursive_children(ids2, model, parent_field)
```
Comment (xmo-odoo) [inline openerp/osv/expression.py]: Could recursive_children return a recordset instead of a list of ids?

File: openerp/osv/expression.py [hunk 3]
```diff
@@ -703,47 +704,51 @@ def to_ids(value, comodel, context=None, limit=None):
...
                 doms = []
-                for node in left_model.browse(cr, uid, ids, context=context):
+                for rec in left_model.browse(ids):
                     if doms:
                         doms.insert(0, OR_OPERATOR)
-                    doms += [AND_OPERATOR, ('parent_right', '>', node.parent_left), ('parent_left', '<=',  node.parent_left)]
+                    doms += [AND_OPERATOR, ('parent_right', '>', rec.parent_left), ('parent_left', '<=',  rec.parent_left)]
                 if prefix:
-                    return [(left, 'in', left_model.search(cr, uid, doms, context=context))]
+                    return [(left, 'in', left_model.search(doms))]
```
Comment (xmo-odoo) [inline openerp/osv/expression.py]: Do domains work with recordset values?

**Comment by xmo-odoo** [comment]
Can't comment inline in `models.py` (diff too large for github) so:

* I'm not fond of aligned inline comments (for BaseModel class fields), if new longer fields are added everything needs to be re-indented or everything looks like garbage. Also doesn't support multiline comments so we end up with both "pre-comments" and "inline comments"
* should `view_init` still be a thing? It seems like a horrible hack and the only place where it's still used is a crm.lead wizard which could check whatever it does in the method/action opening the wizard. Also seems odd to specify exactly when it's called, that feels like an implementation detail
* line 448 could use a set literal:

  ```python
  names = set(vals) - set(['model', 'name'])
  ```
  ->
  ```python
  names = set(vals) - {'model', 'name'}
  ```
* line 1035 the inlined construction of `message` could also use proper string formatting rather than the concatenation of a translated string and a formatted one (and maybe named parameters for translator clarity)
* fields_get_keys should probably be deprecated (or even removed), it's pretty pointless and not used anywhere (was probably a shortcut for a client). Also corresponding comment line 2789
* line 1340… `element` could probably be inlined in the view's call in both search and tree, also the first field in the view in `_get_default_calendar_view`
* line 1710, `_add_missing_default_values` could use set comprehensions
* line 2404, maybe `_init_column` should be a method on `openerp.fields.Field` instead?
* line 2484, is it normal that this still works on _columns rather than _fields?
* lines 2450, 2726: moving `_foreign_keys` from the instance to the field seems a bit odd. Also maybe there should be a fixme somewhere to see if we could replace that hack by init todos in the future?
* line 2894: there's a semantic change here. I assume psycopg allows executing multiple statements, but the old code committed after each SQL statement, the new one only does so at the very end
* line 3538: why introduce `nclauses` instead of just using `params` and `len(params)`?
* line 3879: assuming `check_field_access_rights` will behave correctly when given a dict seems a bit optimistic/dangerous, same for `_store_get_values` line 3891. Since that's also used line 3895, maybe call `vals.keys()`/`list(vals)` just once and reuse it several times?
* also iteration line 3885 could use `iteritems()`
* line 3921: the `self.env.lang` test is not necessary anymore (if it's falsy it's different than `"en_US"` anyway)
* line 4406: `itertools.chain.from_iterable`
* line 4757 (copy_data), maybe `default` should be copied before setting `state` into it? Seems a bit odd to add `state` to the original `default` but copy it before the final conversions
* also why does `copy_data` now return a list?
* line 5199: we know `active_test` is in the context, should use `del context['active_test']` not pop
* line 5207: could use a dictcomp

**Pull request merged**